### ISSUE TITLE
Move decompositions to a separate class.

### ISSUE THREAD

**Issue by Beakerboy** [markrogoyski/math-php#330]
To me, it makes little sense to have $L, $U, and $P as Matrix class properties, especially given that not all matrices HAVE an LUP decomposition. I propose, instead, having a $decompositions array. The results of LU, QR, SVD, Eigen, and others can be saved there for future use.
````php
$decomposions['qr'] = [
    'q' => $Q,
    'r' => $R,
];
$decomposions['power_iteration'] = [
    'q' => $EVec,
    'lambda' => MatrixFactory::diagonal($EVal),
];
$decomposions['jacobi_method'] = [
    'q' => $EVec,
    'lambda' => MatrixFactory::diagonal($EVal),
];
````
EigenValues / EigenVectors are just another form of matrix decomposition, A = QΛQ⁻¹ where Q is the eigenvectors and Λ is the diagonal matrix of eigenvalues. Most Eigen algorithms produce both values and vectors, so it is a waste of time having to recalculate one or the other.

**Comment by Beakerboy** [comment]
I propose something like this for eigenvalues:
````
    /**************************************************************************
     * EIGEN METHODS
     * - eigenvalues
     * - eigenvectors
     **************************************************************************/
    /**
     * Eigenvalues of the matrix.
     * Various eigenvalue algorithms (methods) are available.
     * Use the $method parameter to control the algorithm used.
     *
     * @param string $method Algorithm used to compute the eigenvalues
     *
     * @return array of eigenvalues
     *
     * @throws Exception\MatrixException if method is not a valid eigenvalue method
     */
    public function eigenvalues(string $method = null): array
    {
        if (isset($this->decompositons["eigen"]["D"])) {
            return $this->decompositons["eigen"]["D"];
        }
        if ($method !== null && !EigenDecomposition::isAvailableMethod($method)) {
            throw new Exception\MatrixException("$method is not a valid eigenvalue method");
        }
        if (!$this->isSquare()) {
            throw new Exception\MatrixException('Eigenvalues can only be calculated on square matrices');
        }
        if ($method == EigenDecomposition::CLOSED_FORM_POLYNOMIAL_ROOT_METHOD || ($method === null && $this->m < 5)) {
            $this->decompositons["eigen"] = EigenDecomposition::closedFormPolynomialRootMethod($this);
            return $this->decompositons["eigen"]["D"];
        }
        if ($method == EigenDecomposition::JK_METHOD || ($method === null && $this->isSymmetric())) {
            $this->decompositons["eigen"] = EigenDecomposition::JKMethod($this);
            return $this->decompositons["eigen"]["D"];
        }
        if ($this->isTriangular()) {
            $this->decompositons["eigen"]["D"] = $this->getDiagonalElements();
            $this->decompositons["eigen"]["P"] = EigenDecomposition::eigenvectorsFromEigenvalues($this, $this->decompositons["eigen"]["D"]);
            return $this->decompositons["eigen"]["D"];
        }
        throw new Exception\MatrixException("Eigenvalue cannot be calculated");
    }
    /**
     * Eigenvectors of the matrix.
     * Eigenvector computation function takes in an array of eigenvalues as input.
     * Various eigenvalue algorithms (methods) are availbale.
     * Use the $method parameter to control the algorithm used.
     *
     * @param string $method Algorithm used to compute the eigenvalues
     *
     * @return Matrix of eigenvectors
     *
     * @throws Exception\MatrixException if method is not a valid eigenvalue method
     * @throws Exception\BadDataException
     */
    public function eigenvectors(string $method = null): Matrix
    {
        if (isset($this->decompositons["eigen"]["P"])) {
            return $this->decompositons["eigen"]["P"];
        }
        if ($method === null) {
            $this->eigenvalues();
        } else {
            $this->eigenvalues($method);
        }
        return $this->decompositions["eigen"]["P"];
    }
````

**Comment by markrogoyski** [comment]
Thanks for the suggestions and thinking about this. I have also thought about this and was thinking about making these changes in Version 2 as to not keep adding things to the Version 1 list so something concrete can be released.

Anyways, you are right that the decompositions are taking up a lot of space in the Matrix class and would make sense to be someplace else. I would approach the problem a bit differently though.

1) Each decomposition would be its own class
2) A Matrix would have a Catalog of decompositions
3) Stop using arrays everywhere

**Each decomposition in its own class**
The logic to compute things like QR or LU(P) are quite involved, and is only kind of related to the rest of the Matrix code. A Matrix might have a QR decomposition, and there are multiple ways to compute it. The computation is a detail from the Matrix class's perspective and I would hide that from the Matrix.

Each decomposition class could have a simple static factory method to "create" (i.e., compute) the QR decomposition and construct the QR object. This object would provide access to the Q and R matrices. I'd like the Matrix `qrDecomposition` method to return the QR object, rather than an array of Matrix objects.

**Matrix would have a Catalog of decompositions**
I'm thinking a Matrix would have a Catalog of decompositions and eigenvalues. From within the Matrix you would just pull what you want from the Catalog. If you've already computed it, you get it. If it is the first time you are asking for it, it gets computed (created) at that time and stored in the Catalog. You can think of the Catalog as a smart collection. This way, you don't need any Matrix properties for these things other than the Catalog, which is just a mechanism to lazily load things as they are needed. It contains a catalog of things the Matrix may need, and the Matrix isn't really concerned with where the Catalog got them from.

The nice thing about this is there would be no conditional logic to check if something is in the Catalog. You just ask for it and you get it.

**Stop using arrays everywhere**
Arrays are very convenient in PHP, but they also don't have types and don't provide any information about what might be inside them. I'd prefer return values of Decomposition or Eigenvalue objects and such, and some sort of Collection class (Catalog above) to hold all of these.

This is my current thinking about how to reorganize some of these Matrix-related operations into separate classes in a more OOP way that makes the code easier to reason about and work with.

Thoughts?

**Comment by Beakerboy** [comment]
I like it. I had originally had the Eigen class as one object. I’ll have to read up on Collections.

**Comment by markrogoyski** [comment]
I've moved all the decompositions and reductions into their own classes in the develop branch. The matrix no longer has any properties for any of these derived values. I was thinking I would not be able to return the decomposition classes from the Matrix because it would break the existing return value of an array, but I figured out I could just implement `\ArrayAccess` on those objects and any existing use case would still work. So that worked out nicely.

I've added a simple catalog object that takes care of managing all the derived data. I will probably iterate on it if I decide I want to make it smarter, but as it is it serves its purpose to simplify the Matrix class.

I'll think about how Eigenvalues/vectors fit into this. I still need to add some more unit tests, but in the meantime, I think we can close this issue.
